articles/service-fabric/service-fabric-reliable-services-communication-remoting.md

```diff
@@ -311,50 +301,59 @@ Following example uses Json Serialization with Remoting V2.
             var list = new List<ArraySegment<byte>> { segment };
             return new OutgoingMessageBody(list);
         }
-
+    
         public IServiceRemotingResponseMessageBody Deserialize(IncomingMessageBody messageBody)
         {
             using (var sr = new StreamReader(messageBody.GetReceivedBuffer()))
-
             using (var reader = new JsonTextReader(sr))
             {
                 var serializer = JsonSerializer.Create(new JsonSerializerSettings()
                 {
                     TypeNameHandling = TypeNameHandling.All
                 });
-
+    
                 return serializer.Deserialize<JsonRemotingResponseBody>(reader);
             }
         }
     }
-
-    internal class JsonRemotingResponseBody : IServiceRemotingResponseMessageBody
+    
+    class JsonRemotingResponseBody : IServiceRemotingResponseMessageBody
     {
-        public object Value;
-
+        public string Value { get; set; }
+    
         public void Set(object response)
         {
-            this.Value = response;
+            Value = JsonConvert.SerializeObject(response);
```

Comment:
Honestly, I think the MS team needs to improve this sample if it's that clearly broken.

Reply:
After debugging the code more, we are reminded why the double serialization is there. The `Value` property in the `IServiceRemotingResponseMessageBody` implementation is serialized **without type information** which is then serialized again in `IServiceRemotingResponseMessageBodySerializer` implementation, this time **with the type information**.

The reason for this is that if you use the normal response type to be serialized in `Serialize`, then type information is stored in the JSON created for the `Value` property. The payload is orders of magnitude larger than if it is just a regular JSON `string` because it will include ALL properties with default values and ALL class/value type information. For a complex object and chatty services, you will quickly run out of memory from all the string allocations.

Therefore, serializing the `Value` property without type information and excluding properties with default values creates a much smaller payload for complex objects which is worth the CPU overhead vs memory use. Different people's mileage may vary, but in general we serialize arrays of large objects between services so doing it this way is the better of the two evils.

An optimization would be to check for value types and strings and avoid the extra serialization of `Value` since you get no benefit in turning those into JSON. A much greater optimization would be to statically create the `JsonSerializerSettings` object which is a very expensive operation.

The type failures where in parameter serialization because there is no type conversion on line #357 which I'm guessing breaks the dynamic invocation down the line on the receiver side. I was confused by your comment about the "second serialization" when in fact the `IServiceRemotingResponseMessageBody` value object is serialized first and then serialized again in `Serialize` to maintain type information.

This entire sample could be improved all over the place, our current production code doesn't look much like the sample anyway as we have our own optimizations in place to improve comms specific to our environment.

Reply:
@brutaldev 
I think your issue is with Request type (IServiceRemotingRequestMessageBody) but your change is for the IServiceRemotingResponseMessageBody .
Lets have a quick call to discuss this PR . Could you schedule meeting with suchia@microsoft.com?